src/debug_tools/stats.rs

```diff
@@ -0,0 +1,376 @@
+    pub b: T,
+}
+
+impl<T: Display + ToPrimitive> Display for RatioStat<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{:.3} / {:.3} ({:.3}%)",
+            self.a,
+            self.b,
+            self.get_a_over_b_perc()
+        )
+    }
+}
+
+impl<T: ToPrimitive> RatioStat<T> {
+    /// `new` doesn't have any logic, but this will reduce a lot of line lengths
+    /// since this is called so many times.
+    fn new(a: T, b: T) -> Self {
+        Self { a, b }
+    }
+
+    fn get_a_over_b_perc(&self) -> f32 {
+        (self.a.to_f32().unwrap() / self.b.to_f32().unwrap()) * 100.0
+    }
+}
+
+/// "Raw" state that is mutated as we traverse down the trie. Is processed into
+/// a more useful format later on.
+#[derive(Debug, Default)]
+struct CurrTrackingState {
+    counts: NodeCounts,
+
+    // The "*_sum" variables are just accumulators that we process later to get average depths.
+    leaf_depth_sum: u64,
+    hash_depth_sum: u64,
+    lowest_depth: usize,
+}
+
+impl CurrTrackingState {
+    fn update_lowest_depth_if_larger(&mut self, curr_depth: usize) {
+        if self.lowest_depth < curr_depth {
+            self.lowest_depth = curr_depth;
+        }
+    }
+}
+
+/// Depth in terms of node depth (not key length).
+#[derive(Clone, Debug, Default)]
+pub struct DepthStats {
+    pub lowest_depth: usize,
+    pub avg_leaf_depth: f32,
+    pub avg_hash_depth: f32,
+}
+
+impl Display for DepthStats {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "Lowest depth: {}", self.lowest_depth)?;
+        writeln!(f, "Average leaf depth: {:.3}", self.avg_leaf_depth)?;
+        writeln!(f, "Average hash depth: {:.3}", self.avg_hash_depth)
+    }
+}
+
+impl DepthStats {
+    fn compare(&self, other: &Self) -> DepthComparison {
+        DepthComparison {
+            lowest_depth_rat: RatioStat::new(self.lowest_depth, other.lowest_depth),
+            avg_leaf_depth_rat: RatioStat::new(self.avg_leaf_depth, other.avg_leaf_depth),
+            avg_hash_depth_rat: RatioStat::new(self.avg_hash_depth, other.avg_hash_depth),
+        }
+    }
+}
+
+pub fn get_trie_stats<T: PartialTrie>(trie: &T) -> TrieStats {
+    get_trie_stats_common(trie, None)
+}
+
+pub fn get_trie_stats_with_name<T: PartialTrie>(trie: &T, name: String) -> TrieStats {
+    get_trie_stats_common(trie, Some(name))
+}
+
+fn get_trie_stats_common<T: PartialTrie>(trie: &T, name: Option<String>) -> TrieStats {
+    let mut state = CurrTrackingState::default();
+
+    get_trie_stats_rec(trie, &mut state, 0);
+
+    let depth_stats = DepthStats {
+        lowest_depth: state.lowest_depth,
+        avg_leaf_depth: state.leaf_depth_sum as f32 / state.counts.leaf as f32,
+        avg_hash_depth: state.hash_depth_sum as f32 / state.counts.hash as f32,
+    };
+
+    TrieStats {
+        name,
+        counts: state.counts,
+        depth_stats,
+    }
+}
+
+fn get_trie_stats_rec<T: PartialTrie>(
```

Comment:
returning the state rather than using a mut state as an arg would be more rusty

Reply:
The only downside is then we would need to do more computation when merging together multiple `CurrTrackingState` together, which I guess doesn't really matter too much.

Reply:
Will revisit this afterwards!